xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -842,24 +869,98 @@ static String findClusterNameInRouteConfig(RouteConfiguration config, String hos
       }
     }
 
+    List<EnvoyProtoData.Route> routes = new ArrayList<>();
     // Proceed with the virtual host that has longest wildcard matched domain name with the
     // hostname in original "xds:" URI.
     // Note we would consider upstream cluster not found if the virtual host is not configured
     // correctly for gRPC, even if there exist other virtual hosts with (lower priority)
     // matching domains.
     if (targetVirtualHost != null) {
-      // The client will look only at the last route in the list (the default route),
-      // whose match field must contain a prefix field whose value is empty string
-      // and whose route field must be set.
-      List<Route> routes = targetVirtualHost.getRoutesList();
-      if (!routes.isEmpty()) {
-        Route route = routes.get(routes.size() - 1);
-        if (route.getMatch().getPrefix().isEmpty()) {
-          if (route.hasRoute()) {
-            return route.getRoute().getCluster();
-          }
+      List<Route> routesProto = targetVirtualHost.getRoutesList();
+      for (Route route : routesProto) {
+        routes.add(EnvoyProtoData.Route.fromEnvoyProtoRoute(route));
+      }
+    }
+    return routes;
+  }
+
+  /**
+   * Validates the given list of routes and return error details if there's any error.
+   */
+  // We only validate the default route unless path matching is enabled.
+  // We do more validation if path matching is enabled, but whether every single route should be
+  // valid or any type of error can be skipped is TBD. For now we consider the whole list invalid on
+  // any single error.
+  @Nullable
+  private static String validateRoutes(List<EnvoyProtoData.Route> routes) {
+    if (routes.isEmpty()) {
+      return "No routes found";
+    }
+
+    if (!enablePathMatching) {
+      EnvoyProtoData.Route route = routes.get(routes.size() - 1);
+      RouteMatch routeMatch = route.routeMatch;
+      if (!routeMatch.path.isEmpty() || !routeMatch.prefix.isEmpty() || routeMatch.hasRegex) {
+        return "The last route must be the default route";
+      }
+      if (!route.routeAction.isPresent()) {
+        return "Route action is not specified for the default route";
+      }
+      if (route.routeAction.get().cluster.isEmpty()) {
+        return "Cluster is not specified for the default route";
+      }
+      return null;
+    }
+
+    Set<String> prefixMatches = new HashSet<>();
+    Set<String> pathMatches = new HashSet<>();
+    for (int i = 0; i < routes.size(); i++) {
+      EnvoyProtoData.Route route = routes.get(i);
+
+      if (!route.routeAction.isPresent()) {
+        return "Route action is not specified for one of the routes";
+      }
+
+      RouteMatch routeMatch = route.routeMatch;
+      String prefix = routeMatch.prefix;
+      String path = routeMatch.path;
+      // Whether some of the errors can be skipped is TBD.
```

Comment:
nit: if it is something TBD. Mark it as `TODO`. Also I would suggest rephrasing the sentence .